src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/visitor/AntlrSqlParseTreeVisitor.java

```diff
@@ -0,0 +1,279 @@
+import static com.amazon.opendistroforelasticsearch.sql.antlr.parser.OpenDistroSqlParser.SelectExpressionElementContext;
+import static com.amazon.opendistroforelasticsearch.sql.antlr.parser.OpenDistroSqlParser.SelectFunctionElementContext;
+import static com.amazon.opendistroforelasticsearch.sql.antlr.parser.OpenDistroSqlParser.SimpleTableNameContext;
+import static com.amazon.opendistroforelasticsearch.sql.antlr.parser.OpenDistroSqlParser.StringLiteralContext;
+import static com.amazon.opendistroforelasticsearch.sql.antlr.parser.OpenDistroSqlParser.TableAndTypeNameContext;
+import static com.amazon.opendistroforelasticsearch.sql.antlr.parser.OpenDistroSqlParser.UdfFunctionCallContext;
+import static com.amazon.opendistroforelasticsearch.sql.antlr.parser.OpenDistroSqlParser.UidContext;
+
+/**
+ * ANTLR parse tree visitor to drive the analysis process.
+ */
+public class AntlrSqlParseTreeVisitor<T extends Reducible> extends OpenDistroSqlParserBaseVisitor<T> {
+
+    /** Generic visitor to perform the real action on parse tree */
+    private final GenericSqlParseTreeVisitor<T> visitor;
+
+    public AntlrSqlParseTreeVisitor(GenericSqlParseTreeVisitor<T> visitor) {
+        this.visitor = visitor;
+    }
+
+    @Override
+    public T visitRoot(RootContext ctx) {
+        visitor.visitRoot();
+        return super.visitRoot(ctx);
+    }
+
+    /**
+     * Enforce visit order because ANTLR is generic and unaware.
+     *
+     * Visiting order is:
+     *  FROM
+     *  => WHERE
+     *   => SELECT
+     *    => GROUP BY
+     *     => HAVING
+     *      => ORDER BY
+     *       => LIMIT
+     */
+     @Override
+    public T visitQuerySpecification(QuerySpecificationContext ctx) {
+        visitor.visitQuery();
+
+        // Always visit FROM clause first to define symbols
+        FromClauseContext fromClause = ctx.fromClause();
+        visit(fromClause.tableSources());
+
+        if (fromClause.whereExpr != null) {
+            visit(fromClause.whereExpr);
+        }
+
+        // Note visit GROUP BY and HAVING later than SELECT for alias definition
+        T result = visitSelectElements(ctx.selectElements());
+        fromClause.groupByItem().forEach(this::visit);
+        if (fromClause.havingExpr != null) {
+            visit(fromClause.havingExpr);
+        }
+
+        if (ctx.orderByClause() != null) {
+            visitOrderByClause(ctx.orderByClause());
+        }
+        if (ctx.limitClause() != null) {
+            visitLimitClause(ctx.limitClause());
+        }
+
+        visitor.endVisitQuery();
+        return result;
+    }
+
+    @Override
+    public T visitAtomTableItem(AtomTableItemContext ctx) {
+        String alias = (ctx.alias == null) ? "" : ctx.alias.getText();
+        T result;
+        if (ctx.tableName() instanceof SimpleTableNameContext) {
+            String tableName = ctx.tableName().getText();
+            if (isPath(tableName)) {
```

Comment:
The isPath can't not handle the case if the index name include ".". I am wondering do we need to differentiate nested index when doing semantic analysis?

Reply:
Good point. I wasn't aware of this. Need to figure out because name with dot can mess up the entire analysis if it's allowed in our plugin.

Reply:
It seems many other code will also break. Maybe we should consider how to avoid it later.